File: client/lib/models/responses/exercise_set.dart
/// One set in a live or completed workout. Mirrors backend `ExerciseSetDto`.
/// `pr` is set by the finish-session pass; while a session is in-progress
/// every set's `pr` is false.
class ExerciseSet {
  ExerciseSet({
    required this.id,
    required this.sessionExerciseId,
    required this.setNumber,
    required this.done,
    required this.pr,
    this.weightKg,
    this.reps,
    this.rpe,
    this.completedAt,
  });

  final int id;
  final int sessionExerciseId;
  final int setNumber;
  final double? weightKg;
  final int? reps;
  final double? rpe;
  final bool done;
  final bool pr;
  final DateTime? completedAt;

  factory ExerciseSet.fromJson(Map<String, dynamic> json) {
    return ExerciseSet(
      id: (json['id'] as num).toInt(),
      sessionExerciseId: (json['sessionExerciseId'] as num).toInt(),
      setNumber: (json['setNumber'] as num).toInt(),
      weightKg: (json['weightKg'] as num?)?.toDouble(),
      reps: (json['reps'] as num?)?.toInt(),
      rpe: (json['rpe'] as num?)?.toDouble(),
      done: json['done'] as bool? ?? false,
      pr: json['pr'] as bool? ?? false,
      completedAt: json['completedAt'] == null
          ? null
          : DateTime.tryParse(json['completedAt'] as String),
    );
  }
}

File: api/src/main/java/com/example/athletehub/model/ExerciseSet.java
package com.example.athletehub.model;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.OffsetDateTime;

/**
 * A single set in a live or completed workout. The {@code is_pr} flag is
 * set by the finish-session pass (AH-033) when this set's
 * (weight, reps) produced a new personal-record for its exercise on at
 * least one metric (e1RM or max_weight); kept on the set row so the live
 * workout UI can highlight specific lines without joining back through
 * {@code personal_records}.
 */
@Entity
@Table(name = "exercise_sets")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class ExerciseSet {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "session_exercise_id", nullable = false)
    private Long sessionExerciseId;

    @Column(name = "set_number", nullable = false)
    private int setNumber;

    @Column(name = "weight_kg")
    private BigDecimal weightKg;

    private Integer reps;

    private BigDecimal rpe;

    @Column(name = "is_done", nullable = false)
    private boolean done;

    @Column(name = "is_pr", nullable = false)
    private boolean pr;

    @Column(name = "completed_at")
    private OffsetDateTime completedAt;
}
